ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java

```diff
@@ -2971,6 +2969,87 @@ public void copyConnectionToProject() {
+    SpoonTreeLeveledSelection leveledSelection = (SpoonTreeLeveledSelection) selectionObject;
+    SlaveServerManagementInterface slaveServerManager = null;
+    SlaveServer slaveServer = null;
+    try {
+      slaveServerManager = getSlaveServerManager( leveledSelection.getLevel() );
+      slaveServer = slaveServerManager.get( leveledSelection.getName() );
+      getLog().logBasic( "Moving the slave server " + leveledSelection.getName() + " to global" );
+      delegates.slaves.moveToGlobal( slaveServerManager, slaveServer );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, "Error", "Unexpected error retrieving slave servers", e );
+    }
+  }
+
+  /**
+   * Moving the slaveServer to project
+   */
+  public void moveSlaveServerToProject() {
+    SpoonTreeLeveledSelection leveledSelection = (SpoonTreeLeveledSelection) selectionObject;
+    SlaveServerManagementInterface slaveServerManager = null;
+    SlaveServer slaveServer = null;
+    try {
+      slaveServerManager = getSlaveServerManager( leveledSelection.getLevel() );
+      slaveServer = slaveServerManager.get( leveledSelection.getName() );
+      getLog().logBasic( "Moving the slave server " + leveledSelection.getName() + " to project" );
+      delegates.slaves.moveToProject( slaveServerManager, slaveServer );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, "Error", "Unexpected error retrieving slave servers", e );
+    }
+  }
+
+  /**
+   * Copy the slaveServer to global level
+   */
+  public void copySlaveServerToGlobal() {
+    SpoonTreeLeveledSelection leveledSelection = (SpoonTreeLeveledSelection) selectionObject;
+    SlaveServerManagementInterface slaveServerManager = null;
+    SlaveServer slaveServer = null;
+    try {
+      slaveServerManager = getSlaveServerManager( leveledSelection.getLevel() );
+      slaveServer = slaveServerManager.get( leveledSelection.getName() );
+      getLog().logBasic( "Moving the slave server " + leveledSelection.getName() + " to global" );
+      delegates.slaves.copyToGlobal( slaveServerManager, slaveServer );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, "Error", "Unexpected error retrieving slave servers", e );
+    }
+  }
+
+  /**
+   * Copy the slaveServer to project
+   */
+  public void copySlaveServerToProject() {
+    SpoonTreeLeveledSelection leveledSelection = (SpoonTreeLeveledSelection) selectionObject;
+    SlaveServerManagementInterface slaveServerManager = null;
+    SlaveServer slaveServer = null;
+    try {
+      slaveServerManager = getSlaveServerManager( leveledSelection.getLevel() );
+      slaveServer = slaveServerManager.get( leveledSelection.getName() );
+      getLog().logBasic( "Moving the slave server " + leveledSelection.getName() + " to project" );
+      delegates.slaves.copyToProject( slaveServerManager, slaveServer );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, "Error", "Unexpected error retrieving slave servers", e );
+    }
+  }
```

Comment:
this is a lot of copy-paste code. Can we combine these?

Reply:
I did look at the different options. I could move the logic to delegate so the delegate method would be
 delegates.slaves.copyToProject( SpoonTreeLeveledSelection leveledSelection );
But I am just moving the code from Spoon to delegate class, not reusing it. 
If we combine, then we would need a big switch statement to calls different method on delegate based on parameter.

Reply:
Could do something like: 
```
  @FunctionalInterface
  private static interface SlaveServerOp {
    void op( SlaveServerManager ssm, slaveServer ss ) throws SomethingUsefulException;
  }

  private void withSlaveServer( String message, SlaveServerOp sso ) {
    SpoonTreeLeveledSelection leveledSelection = (SpoonTreeLeveledSelection) selectionObject;
    SlaveServerManagementInterface slaveServerManager = null;
    SlaveServer slaveServer = null;
    try {
      slaveServerManager = getSlaveServerManager( leveledSelection.getLevel() );
      slaveServer = slaveServerManager.get( leveledSelection.getName() );
      getLog().logBasic(  );
      sso.op( slaveServerManager, slaveServer );
    } catch ( Exception e ) {
      new ErrorDialog( shell, "Error", "Unexpected error retrieving slave servers", e );
    }
  }

  /**
   * Copy the slaveServer to project
   */
  public void copySlaveServerToProject() {
    withSlaveServer( "Copying the slave server " + leveledSelection.getName() + " to project", 
      ( ssm, ss ) -> delegates.slaves.copyToProject( ssm, ss ) );
  }
```

NOTE: the log messages are wrong for copy, too.